src/notebook/notebookFeature.ts

```diff
@@ -49,11 +69,39 @@ export class JuliaNotebookFeature {
                 }
             })
             controller.executeHandler = this.executeCells.bind(this)
+            controller.onDidChangeSelectedNotebooks(({ notebook, selected }) => {
+                // If we select our controller, then update the notebook metadata with the kernel information.
+                if (!selected) {
+                    return;
+                }
+                this.updateNotebookWithSelectedKernel(notebook, displayName, ver);
+            }, this, this.disposables)
 
-            this._controllers.push(controller)
+            this._controllers.set(controller, {version: ver })
         }
     }
 
+    private onDidOpenNotebookDocument(e: vscode.NotebookDocument) {
+        if (!this.isJuliaNotebook(e) || this._controllers.size === 0) {
+            return;
+        }
+        // Get metadata from notebook (to get an hint of what version of julia is used)
+        const { name, version } = this.getKernelSpecNameAndVersion(e);
+        let preferredControllerFound = false;
+        this._controllers.forEach((info, controller) => {
+            // If we find a controller that matches the vesion in the notebook metadata, then set
+            // that controller as the preferred controller.
+            if (name.includes(info.version) || version.includes(info.version)) {
```

Comment:
I think we probably want to make the first criterion for being a preferred controller whether the `kernelspec.name` matches the `kernelId` of the controller. That string will have the form `julia-1.6`, i.e. it will include the major and minor version part, but not the patch version. If we don't find a match on that, I think it is not clear to me whether we should set a preferred kernel at all. I think JupyterLab doesn't, for example...

The existing code will always pick the latest patch release in a given `major.minor` combo as the controller, if multiple Julia versions in a given `major.minor` series are installed, which matches the Jupyter experience. I _think_ that makes sense, but it does mean that say a notebook was saved with a Julia 1.6.0 kernel, and then is opened on a system that has Julia 1.6.0 and 1.6.1 installed, a kernel with 1.6.1 is going to be used. I'd say we can try that for now, I think it is what folks expect based on the existing Jupyter implementation.